File: mcp-gateway/ratelimit.py
"""Rate limiting for MCP Gateway — Valkey-backed sliding-window quotas.

Implements:
- Sliding window per-identity/tool rate limiting (ZSET-based)
- Per-IP rate limiting (blocks credential stuffing / distributed abuse)
- Concurrent request limits per-identity
- Per-team RPM overrides (from config bundle)
- Configurable fail mode: fail-open (default) or fail-closed

Exceed → JSON-RPC error -32029 "rate_limited".
"""
import logging
import os
import time
from typing import Optional

logger = logging.getLogger(__name__)

MCP_RATE_LIMITED = -32029

_client = None

# Default config: fail-open for backward compat, configurable via env
_FAIL_CLOSED = os.environ.get("MCP_RATELIMIT_FAIL_CLOSED", "false").lower() in ("true", "1", "yes")
_DEFAULT_MAX_IP_RPM = int(os.environ.get("MCP_MAX_IP_RPM", "0"))  # 0 = disabled
_DEFAULT_MAX_CONCURRENT = int(os.environ.get("MCP_MAX_CONCURRENT", "0"))  # 0 = disabled
_WINDOW_SECONDS = 60


def _get_client():
    """Return a Valkey client, or None if unavailable."""
    global _client
    if _client is not None:
        return _client
    try:
        from valkey import Valkey
        host = os.environ.get("VALKEY_HOST", "valkey")
        port = int(os.environ.get("VALKEY_PORT", "6379"))
        password = os.environ.get("VALKEY_PASSWORD") or None
        client = Valkey(
            host=host, port=port, db=0, password=password,
            socket_connect_timeout=1, socket_timeout=1,
            decode_responses=True,
        )
        client.ping()
        _client = client
        return client
    except Exception as e:
        logger.debug("Valkey not available: %s", e)
        _client = None
        return None


def _sliding_window_check(
    client,
    key: str,
    max_count: int,
    window: int = _WINDOW_SECONDS,
) -> tuple[bool, int]:
    """Sliding window check using ZSET. Returns (allowed, remaining).

    Removes entries older than window, counts remaining, adds current timestamp.
    """
    now = time.time()
    cutoff = now - window

    pipe = client.pipeline()
    pipe.zremrangebyscore(key, 0, cutoff)
    pipe.zcard(key)
    _, count = pipe.execute()

    if count >= max_count:
        return False, 0

    member = f"{now}:{count}"
    pipe = client.pipeline()
    pipe.zadd(key, {member: now})
    pipe.expire(key, window + 5)
    pipe.execute()

    return True, max_count - count - 1


def check_rate_limit(
    identity_id: int,
    tool: str,
    max_rpm: int,
) -> tuple[bool, int]:
    """Check if a request is within the sliding-window rate limit.

    Returns (allowed, remaining). Uses a 60-second sliding window via ZSET.
    Key: mcp:rl:{identity_id}:{tool}
    """
    client = _get_client()
    if not client:
        if _FAIL_CLOSED:
            logger.warning("Rate limit check failed (fail-closed: denying)")
            return False, 0
        return True, max_rpm

    if max_rpm <= 0:
        return True, max_rpm

    key = f"mcp:rl:{identity_id}:{tool}"
    try:
        return _sliding_window_check(client, key, max_rpm)
    except Exception as e:
        if _FAIL_CLOSED:
            logger.warning("Rate limit check failed (fail-closed: denying): %s", e)
            return False, 0
        logger.warning("Rate limit check failed (allowing): %s", e)
        return True, max_rpm


def check_ip_rate_limit(
    ip: str,
    max_ip_rpm: Optional[int] = None,
) -> tuple[bool, int]:
    """Check per-IP rate limit. Returns (allowed, remaining).

    Disabled if max_ip_rpm is 0 or None.
    """
    limit = max_ip_rpm if max_ip_rpm is not None else _DEFAULT_MAX_IP_RPM
    if limit <= 0:
        return True, limit

    client = _get_client()
    if not client:
        if _FAIL_CLOSED:
            return False, 0
        return True, limit

    key = f"mcp:rl:ip:{ip}"
    try:
        return _sliding_window_check(client, key, limit)
    except Exception as e:
        if _FAIL_CLOSED:
            logger.warning("IP rate limit check failed (fail-closed: denying): %s", e)
            return False, 0
        return True, limit


def acquire_concurrent_slot(
    identity_id: int,
    max_concurrent: Optional[int] = None,
) -> bool:
    """Acquire a concurrent request slot. Returns True if allowed.

    Uses INCR with TTL as a lease. Must be released with release_concurrent_slot.
    Disabled if max_concurrent is 0 or None.
    """
    limit = max_concurrent if max_concurrent is not None else _DEFAULT_MAX_CONCURRENT
    if limit <= 0:
        return True

    client = _get_client()
    if not client:
        if _FAIL_CLOSED:
            return False
        return True

    key = f"mcp:conc:{identity_id}"
    try:
        count = client.incr(key)
        if count == 1:
            client.expire(key, 300)  # 5-min safety expiry
        if count > limit:
            client.decr(key)
            return False
        return True
    except Exception as e:
        if _FAIL_CLOSED:
            logger.warning("Concurrent check failed (fail-closed: denying): %s", e)
            return False
        return True


def release_concurrent_slot(identity_id: int) -> None:
    """Release a concurrent request slot."""
    if _DEFAULT_MAX_CONCURRENT <= 0:
        return
    client = _get_client()
    if not client:
        return
    key = f"mcp:conc:{identity_id}"
    try:
        client.decr(key)
    except Exception:
        pass


def get_team_rpm(config: dict, team_id: int, default_rpm: int) -> int:
    """Get the RPM for a team, falling back to default_rpm.

    Config bundle may include 'team_rpm_overrides': {team_id: rpm}.
    """
    overrides = config.get("team_rpm_overrides") or {}
    return int(overrides.get(str(team_id), overrides.get(team_id, default_rpm)))


def get_rate_limit_remaining(identity_id: int, tool: str, max_rpm: int) -> int:
    """Return remaining requests in the current sliding window."""
    client = _get_client()
    if not client or max_rpm <= 0:
        return max_rpm
    key = f"mcp:rl:{identity_id}:{tool}"
    try:
        now = time.time()
        cutoff = now - _WINDOW_SECONDS
        client.zremrangebyscore(key, 0, cutoff)
        count = client.zcard(key)
        return max(0, max_rpm - count)
    except Exception:
        return max_rpm

File: mcp-gateway-rs/crates/policy/src/ratelimit.rs
//! Rate limiting — Valkey-backed sliding-window quotas, `mcp:gw:` prefixed.
//!
//! Mirrors `mcp-gateway/ratelimit.py`:
//! - Per-identity/tool sliding window (ZSET): `mcp:gw:rl:{id}:{tool}`
//! - Per-IP sliding window: `mcp:gw:rl:ip:{ip}`
//! - Concurrent request leases (INCR + TTL): `mcp:gw:conc:{id}`
//! - Per-team RPM overrides from the config bundle
//! - Configurable fail mode: fail-open (default) or fail-closed
//!
//! Exceed → JSON-RPC error -32029 "rate_limited".

use crate::valkey::ValkeyClient;

pub const MCP_RATE_LIMITED: i32 = -32029;
const WINDOW_SECONDS: u64 = 60;
const CONCURRENT_SAFETY_TTL: u64 = 300;

/// Configuration for the rate limiter.
#[derive(Debug, Clone)]
pub struct RateLimitConfig {
    pub fail_closed: bool,
    pub default_max_ip_rpm: i64,
    pub default_max_concurrent: i64,
}

impl Default for RateLimitConfig {
    fn default() -> Self {
        Self {
            fail_closed: std::env::var("MCP_RATELIMIT_FAIL_CLOSED")
                .map(|v| matches!(v.to_lowercase().as_str(), "true" | "1" | "yes"))
                .unwrap_or(false),
            default_max_ip_rpm: std::env::var("MCP_MAX_IP_RPM")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(0),
            default_max_concurrent: std::env::var("MCP_MAX_CONCURRENT")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(0),
        }
    }
}

#[derive(Clone)]
pub struct RateLimiter {
    client: Option<ValkeyClient>,
    config: RateLimitConfig,
}

/// Result of a rate-limit check.
#[derive(Debug, Clone, Copy)]
pub struct RateLimitDecision {
    pub allowed: bool,
    pub remaining: i64,
}

impl RateLimiter {
    pub fn new(client: Option<ValkeyClient>, config: RateLimitConfig) -> Self {
        Self { client, config }
    }

    /// Check if a request is within the per-identity/tool sliding window.
    pub async fn check_rate_limit(&self, identity_id: i64, tool: &str, max_rpm: i64) -> RateLimitDecision {
        if max_rpm <= 0 {
            return RateLimitDecision { allowed: true, remaining: max_rpm };
        }
        let Some(c) = &self.client else {
            return self.fail_open(max_rpm);
        };
        match c.sliding_window(&format!("rl:{identity_id}:{tool}"), max_rpm, WINDOW_SECONDS).await {
            Ok((allowed, remaining)) => RateLimitDecision { allowed, remaining },
            Err(e) => {
                ValkeyClient::log_failure("rate_limit", &e);
                self.fail_open(max_rpm)
            }
        }
    }

    /// Check per-IP rate limit. Disabled if `max_ip_rpm` is 0.
    pub async fn check_ip_rate_limit(&self, ip: &str, max_ip_rpm: Option<i64>) -> RateLimitDecision {
        let limit = max_ip_rpm.unwrap_or(self.config.default_max_ip_rpm);
        if limit <= 0 {
            return RateLimitDecision { allowed: true, remaining: limit };
        }
        let Some(c) = &self.client else {
            return self.fail_open(limit);
        };
        match c.sliding_window(&format!("rl:ip:{ip}"), limit, WINDOW_SECONDS).await {
            Ok((allowed, remaining)) => RateLimitDecision { allowed, remaining },
            Err(e) => {
                ValkeyClient::log_failure("ip_rate_limit", &e);
                self.fail_open(limit)
            }
        }
    }

    /// Acquire a concurrent request slot. Returns true if allowed.
    /// Must be released with `release_concurrent_slot`.
    pub async fn acquire_concurrent_slot(&self, identity_id: i64, max_concurrent: Option<i64>) -> bool {
        let limit = max_concurrent.unwrap_or(self.config.default_max_concurrent);
        if limit <= 0 {
            return true;
        }
        let Some(c) = &self.client else {
            return !self.config.fail_closed;
        };
        let key = format!("conc:{identity_id}");
        match c.incr(&key).await {
            Ok(count) => {
                if count == 1 {
                    let _ = c.expire(&key, CONCURRENT_SAFETY_TTL).await;
                }
                if count > limit {
                    let _ = c.decr(&key).await;
                    return false;
                }
                true
            }
            Err(e) => {
                ValkeyClient::log_failure("concurrent", &e);
                !self.config.fail_closed
            }
        }
    }

    /// Release a concurrent request slot.
    ///
    /// Always decrements if a Valkey client is present.  The guard in
    /// `acquire_concurrent_slot` already short-circuits when the limit is 0,
    /// so this function is only reached when a slot was actually acquired
    /// (counter >= 1), making the DECR safe.  We must NOT check
    /// `self.config.default_max_concurrent` here because the per-request limit
    /// may come from the config bundle (which can be non-zero even when the
    /// env-var default is zero); checking the env-var default would skip the
    /// DECR and leak the slot, permanently blocking the identity.
    pub async fn release_concurrent_slot(&self, identity_id: i64) {
        if let Some(c) = &self.client {
            let _ = c.decr(&format!("conc:{identity_id}")).await;
        }
    }

    fn fail_open(&self, max_rpm: i64) -> RateLimitDecision {
        if self.config.fail_closed {
            RateLimitDecision { allowed: false, remaining: 0 }
        } else {
            RateLimitDecision { allowed: true, remaining: max_rpm }
        }
    }
}

/// Get the RPM for a team, falling back to `default_rpm`.
/// The config bundle may include `team_rpm_overrides: {team_id: rpm}`.
pub fn get_team_rpm(overrides: &serde_json::Value, team_id: i64, default_rpm: i64) -> i64 {
    if let Some(map) = overrides.as_object() {
        let key = team_id.to_string();
        if let Some(v) = map.get(&key).or_else(|| map.get(&team_id.to_string())) {
            if let Some(n) = v.as_i64() {
                return n;
            }
        }
    }
    default_rpm
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn no_valkey_fail_open() {
        let rl = RateLimiter::new(None, RateLimitConfig::default());
        let d = rl.check_rate_limit(1, "tool", 100).await;
        assert!(d.allowed);
        let d = rl.check_ip_rate_limit("1.2.3.4", Some(100)).await;
        assert!(d.allowed);
        assert!(rl.acquire_concurrent_slot(1, Some(10)).await);
    }

    #[tokio::test]
    async fn no_valkey_fail_closed() {
        let rl = RateLimiter::new(None, RateLimitConfig {
            fail_closed: true,
            default_max_ip_rpm: 100,
            default_max_concurrent: 10,
        });
        let d = rl.check_rate_limit(1, "tool", 100).await;
        assert!(!d.allowed);
        assert!(!rl.acquire_concurrent_slot(1, Some(10)).await);
    }

    #[tokio::test]
    async fn zero_limit_disabled() {
        let rl = RateLimiter::new(None, RateLimitConfig::default());
        let d = rl.check_rate_limit(1, "tool", 0).await;
        assert!(d.allowed);
        let d = rl.check_ip_rate_limit("1.2.3.4", Some(0)).await;
        assert!(d.allowed);
        assert!(rl.acquire_concurrent_slot(1, Some(0)).await);
    }

    #[test]
    fn team_rpm_override() {
        let overrides = serde_json::json!({"1": 200, "2": 50});
        assert_eq!(get_team_rpm(&overrides, 1, 100), 200);
        assert_eq!(get_team_rpm(&overrides, 3, 100), 100);
        assert_eq!(get_team_rpm(&serde_json::json!(null), 1, 100), 100);
    }

    /// Regression test: release_concurrent_slot must always DECR when a client
    /// is present, even if the env-var default (default_max_concurrent) is 0.
    /// The per-request limit may come from the config bundle and be non-zero,
    /// so checking the env-var default in release would leak the slot.
    #[tokio::test]
    async fn release_always_decrements_regardless_of_env_default() {
        // Simulate: env var MCP_MAX_CONCURRENT=0 (default), but config bundle
        // has concurrent_limit=1 (set via UI settings table).
        let rl = RateLimiter::new(None, RateLimitConfig {
            fail_closed: false,
            default_max_ip_rpm: 0,
            default_max_concurrent: 0, // env-var default is 0
        });

        // Without Valkey, acquire returns true (fail-open).  Release is a
        // no-op (no client).  This test documents the contract: release
        // does NOT check default_max_concurrent.
        assert!(rl.acquire_concurrent_slot(1, Some(1)).await);
        // Should not panic or short-circuit — no client so it's a no-op,
        // but the guard must not prevent the DECR path when a client exists.
        rl.release_concurrent_slot(1).await;
    }
}
